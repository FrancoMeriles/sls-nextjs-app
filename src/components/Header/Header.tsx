import React from 'react'
import Image from 'next/image'
import Link from 'next/link'
import { useRouter } from 'next/router'
import {
  Box,
  Flex,
  Badge,
  useDisclosure,
  AlertDialog,
  AlertDialogBody,
  AlertDialogContent,
  AlertDialogOverlay,
} from '@chakra-ui/react'

// Contexts
import { useFavorite } from '@base/contexts/Favorite'

// Icons
import { HeartOutline } from '@base/icons'
import { SearchIcon } from '@chakra-ui/icons'

// Components
import Search from '@base/components/Search'
import ZipCode from '@base/components/ZipCode'

const Header = () => {
  const favorites = useFavorite()
  const router = useRouter()

  const gotoFavorite = () => {
    router.push('/favorites')
  }

  const { isOpen, onOpen, onClose } = useDisclosure()
  const showSearch = () => (isOpen ? onClose() : onOpen())
  const cancelRef = React.useRef()
  return (
    <Flex
      as="nav"
      align="center"
      justify="space-between"
      wrap="wrap"
      padding={6}
      bg="brand.100"
      color="black.300"
    >
      <Flex alignItems="center">
        <Link href="/">
          <a>
            <Image src="/assets/img/logoMeli.png" alt="Mercado Libre" width="134" height="34" />
          </a>
        </Link>
        <ZipCode />
      </Flex>

      <Box display={{ base: 'none', md: 'block' }}>
        <Search onClose={onClose} />
      </Box>
      <Flex>
        <Box
          data-testid="show-search"
          display={{ base: 'block', md: 'none' }}
          cursor="pointer"
          onClick={showSearch}
          mr={5}
        >
          <SearchIcon boxSize={5} />
        </Box>
        <Box pos="relative" cursor="pointer" onClick={gotoFavorite} data-testid="go-to-favorite">
          <HeartOutline boxSize={5} />
          <Badge colorScheme="red" borderRadius="50px" pos="absolute" right="-11" top="-2">
            {favorites.length}
          </Badge>
        </Box>
      </Flex>
      <AlertDialog isOpen={isOpen} leastDestructiveRef={cancelRef} onClose={onClose}>
        <AlertDialogOverlay>
          <AlertDialogContent>
            <AlertDialogBody>
              <Search onClose={onClose} />
            </AlertDialogBody>
          </AlertDialogContent>
        </AlertDialogOverlay>
      </AlertDialog>
    </Flex>
  )
}

export default Header
