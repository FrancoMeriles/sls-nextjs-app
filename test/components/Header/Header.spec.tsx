import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import Header from '@base/components/Header'

const mockPushRouter = jest.fn()

jest.mock('@base/contexts/Favorite', () => ({
  useFavorite: () => [],
}))

jest.mock('@base/components/ZipCode', () => () => <div />)

jest.mock('next/router', () => ({
  useRouter: () => ({
    push: mockPushRouter,
  }),
}))
describe('Header', () => {
  it('should render and match snapshot with required props', () => {
    const tree = renderer.create(<Header />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should redirect to favorite when user click on favorite icon', async () => {
    render(<Header />)
    userEvent.click(screen.getByTestId('go-to-favorite'))
    expect(mockPushRouter).toHaveBeenCalled()
    expect(mockPushRouter).toHaveBeenCalledWith('/favorites')
  })
})
