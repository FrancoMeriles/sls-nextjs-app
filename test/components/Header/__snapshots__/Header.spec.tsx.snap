// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render and match snapshot with required props 1`] = `
<nav
  className="css-kwvze0"
>
  <div
    className="css-70qvj9"
  >
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "inline-block",
            "margin": 0,
            "maxWidth": "100%",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "block",
              "maxWidth": "100%",
            }
          }
        >
          <img
            alt=""
            aria-hidden={true}
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM0IiBoZWlnaHQ9IjM0IiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
            style={
              Object {
                "border": "none",
                "display": "block",
                "margin": 0,
                "maxWidth": "100%",
                "padding": 0,
              }
            }
          />
        </div>
        <noscript>
          <img
            alt="Mercado Libre"
            decoding="async"
            src="/assets/img/logoMeli.png"
            srcSet="/_next/image?url=%2Fassets%2Fimg%2FlogoMeli.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fimg%2FlogoMeli.png&w=384&q=75 2x"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </noscript>
        <img
          alt="Mercado Libre"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </div>
    </a>
    <div />
  </div>
  <div
    className="css-11nn4uy"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="chakra-input__group css-1idadyg"
      >
        <input
          className="chakra-input css-cxmp86"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Buscar"
          value=""
        />
        <div
          className="chakra-input__right-element css-1ctuf01"
        >
          <button
            className="chakra-button css-1a05ek7"
            type="submit"
          >
            <svg
              className="chakra-icon css-1ivt62w"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </form>
  </div>
  <div
    className="css-k008qs"
  >
    <div
      className="css-xv2kcw"
      onClick={[Function]}
    >
      <svg
        className="chakra-icon css-w31a5m"
        focusable={false}
        viewBox="0 0 24 24"
      >
        <path
          d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      className="css-dfpqc0"
      data-testid="go-to-favorite"
      onClick={[Function]}
    >
      <svg
        className="chakra-icon css-w31a5m"
        focusable={false}
        viewBox="0 -28 512 512"
      >
        <path
          d="M256 455.516c-7.29 0-14.316-2.641-19.793-7.438-20.684-18.086-40.625-35.082-58.219-50.074l-.09-.078c-51.582-43.957-96.125-81.918-127.117-119.313C16.137 236.81 0 197.172 0 153.871c0-42.07 14.426-80.883 40.617-109.293C67.121 15.832 103.488 0 143.031 0c29.555 0 56.621 9.344 80.446 27.77C235.5 37.07 246.398 48.453 256 61.73c9.605-13.277 20.5-24.66 32.527-33.96C312.352 9.344 339.418 0 368.973 0c39.539 0 75.91 15.832 102.414 44.578C497.578 72.988 512 111.801 512 153.871c0 43.3-16.133 82.938-50.777 124.738-30.993 37.399-75.532 75.356-127.106 119.309-17.625 15.016-37.597 32.039-58.328 50.168a30.046 30.046 0 01-19.789 7.43zM143.031 29.992c-31.066 0-59.605 12.399-80.367 34.914-21.07 22.856-32.676 54.45-32.676 88.965 0 36.418 13.535 68.988 43.883 105.606 29.332 35.394 72.961 72.574 123.477 115.625l.093.078c17.66 15.05 37.68 32.113 58.516 50.332 20.961-18.254 41.012-35.344 58.707-50.418 50.512-43.051 94.137-80.223 123.469-115.617 30.344-36.618 43.879-69.188 43.879-105.606 0-34.516-11.606-66.11-32.676-88.965-20.758-22.515-49.3-34.914-80.363-34.914-22.758 0-43.653 7.235-62.102 21.5-16.441 12.719-27.894 28.797-34.61 40.047-3.452 5.785-9.53 9.238-16.261 9.238s-12.809-3.453-16.262-9.238c-6.71-11.25-18.164-27.328-34.61-40.047-18.448-14.265-39.343-21.5-62.097-21.5zm0 0"
          fill="currentColor"
        />
      </svg>
      <span
        className="chakra-badge css-1ravyga"
      >
        0
      </span>
    </div>
  </div>
</nav>
`;
